plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java

```diff
@@ -0,0 +1,653 @@
+            assertQueryFails(
+                    format("INSERT INTO %s VALUES (DATE '1969-12-31')", table.getName()),
+                    "Date must be between 1970-01-01 and 2106-02-07 in ClickHouse: 1969-12-31");
+            assertQueryFails(
+                    format("INSERT INTO %s VALUES (DATE '2106-02-08')", table.getName()),
+                    "Date must be between 1970-01-01 and 2106-02-07 in ClickHouse: 2106-02-08");
+        }
+    }
+
+    @Test(dataProvider = "sessionZonesDataProvider")
+    public void testClickHouseDateTime(ZoneId sessionZone)
+    {
+        Session session = Session.builder(getSession())
+                .setTimeZoneKey(TimeZoneKey.getTimeZoneKey(sessionZone.getId()))
+                .build();
+
+        SqlDataTypeTest.create()
+                .addRoundTrip("datetime", "timestamp '1986-01-01 00:13:07'", createTimestampType(0), "TIMESTAMP '1986-01-01 00:13:07'") // time gap in Kathmandu
+                .addRoundTrip("datetime", "timestamp '2018-03-25 03:17:17'", createTimestampType(0), "TIMESTAMP '2018-03-25 03:17:17'") // time gap in Vilnius
+                .addRoundTrip("datetime", "timestamp '2018-10-28 01:33:17'", createTimestampType(0), "TIMESTAMP '2018-10-28 01:33:17'") // time doubled in JVM zone
+                .addRoundTrip("datetime", "timestamp '2018-10-28 03:33:33'", createTimestampType(0), "TIMESTAMP '2018-10-28 03:33:33'") // time double in Vilnius
+                .execute(getQueryRunner(), session, clickhouseCreateTrinoInsert("tpch.test_datetime"));
+
+        addTimestampRoundTrips("timestamp")
+                .execute(getQueryRunner(), session, clickhouseCreateAndInsert("tpch.test_timestamp"));
+        addTimestampRoundTrips("datetime")
+                .execute(getQueryRunner(), session, clickhouseCreateAndInsert("tpch.test_datetime"));
+        addTimestampRoundTrips("datetime32")
+                .execute(getQueryRunner(), session, clickhouseCreateAndInsert("tpch.test_datetime32"));
+    }
+
+    private SqlDataTypeTest addTimestampRoundTrips(String inputType)
+    {
+        return SqlDataTypeTest.create()
+                .addRoundTrip(inputType, "'1986-01-01 00:13:07'", createTimestampType(0), "TIMESTAMP '1986-01-01 00:13:07'") // time gap in Kathmandu
+                .addRoundTrip(inputType, "'2018-03-25 03:17:17'", createTimestampType(0), "TIMESTAMP '2018-03-25 03:17:17'") // time gap in Vilnius
+                .addRoundTrip(inputType, "'2018-10-28 01:33:17'", createTimestampType(0), "TIMESTAMP '2018-10-28 01:33:17'") // time doubled in JVM zone
+                .addRoundTrip(inputType, "'2018-10-28 03:33:33'", createTimestampType(0), "TIMESTAMP '2018-10-28 03:33:33'") // time double in Vilnius
+                .addRoundTrip(format("Nullable(%s)", inputType), "NULL", createTimestampType(0), "CAST(NULL AS TIMESTAMP(0))");
+    }
+
+    @Test(dataProvider = "clickHouseDateTimeMinMaxValuesDataProvider")
+    public void testClickHouseDateTimeMinMaxValues(String timestamp)
```

Comment:
I may need to merge more related code into the commit "Refactor ClickHouse type mapping tests".
e.g. `testClickHouseDateTimeMinMaxValues`,`testUnsupportedClickHouseDateTime`,`clickHouseDateTimeMinMaxValuesDataProvider ` etc.